File: Java/Tests/jdi-uitest-webtests/src/test/java/com/epam/jdi/uitests/testing/unittests/tests/complex/table/SearchRowsColumnsTests.java
package com.epam.jdi.uitests.testing.unittests.tests.complex.table;

import com.epam.commons.map.MapArray;
import com.epam.jdi.uitests.testing.unittests.tests.complex.table.base.SupportTableTestsBase;
import com.epam.jdi.uitests.web.selenium.elements.complex.table.interfaces.ICell;
import com.epam.web.matcher.testng.Check;
import org.testng.annotations.Test;

import java.util.Arrays;

import static com.epam.commons.PrintUtils.print;
import static com.epam.jdi.uitests.web.selenium.elements.complex.table.Column.column;
import static com.epam.jdi.uitests.web.selenium.elements.complex.table.Row.row;
import static com.epam.web.matcher.testng.Assert.areEquals;
import static java.lang.String.format;

/**
 * Created by Natalia_Grebenshchikova on 10/21/2015.
 */
public class SearchRowsColumnsTests extends SupportTableTestsBase {

    private static final String expectedColumn =
            "1:Selenium, Custom, " +
                    "2:TestNG, JUnit, Custom, " +
                    "3:TestNG, JUnit, Custom, " +
                    "4:Log4J, TestNG log, Custom, " +
                    "5:Jenkins, Allure, Custom, " +
                    "6:Custom";
    private final static String expectedRow =
            "Type:Asserter, " +
                    "Now:TestNG, JUnit, Custom, " +
                    "Plans:MSTest, NUnit, Epam";

    @Test
    public void columnByNumTest() {
        MapArray<String, ICell> column = table().column(2);
        areEquals(print(column.select(
                        (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))),
                "1:Selenium, Custom, " +
                        "2:TestNG, JUnit, Custom, " +
                        "3:TestNG, JUnit, Custom, " +
                        "4:Log4J, TestNG log, Custom, " +
                        "5:Jenkins, Allure, Custom, " +
                        "6:Custom");
    }

    @Test
    public void columnByNameTest() {
        MapArray<String, ICell> column = table().column("Now");
        areEquals(print(column.select(
                        (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))),
                "1:Selenium, Custom, " +
                        "2:TestNG, JUnit, Custom, " +
                        "3:TestNG, JUnit, Custom, " +
                        "4:Log4J, TestNG log, Custom, " +
                        "5:Jenkins, Allure, Custom, " +
                        "6:Custom");
    }

    @Test
    public void rowByNumTest() {
        MapArray<String, ICell> row = table().row(2);
        areEquals(print(row.select(
                        (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))),
                "Type:Test Runner, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam");
    }

    @Test
    public void rowByNameTest() {
        MapArray<String, ICell> row = table().row("2");
        areEquals(print(row.select(
                        (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))),
                "Type:Test Runner, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam");
    }

    @Test
    public void columnByCriteriaIntTest() {
        MapArray<String, ICell> column = table().column("TestNG, JUnit, Custom", row(3));
        areEquals(print(column.select(
                        (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))),
                "1:Selenium, Custom, " +
                        "2:TestNG, JUnit, Custom, " +
                        "3:TestNG, JUnit, Custom, " +
                        "4:Log4J, TestNG log, Custom, " +
                        "5:Jenkins, Allure, Custom, " +
                        "6:Custom");
    }

    @Test
    public void columnByCriteriaStringTest() {
        MapArray<String, ICell> column = table().column("TestNG, JUnit, Custom", row("3"));
        areEquals(print(column.select(
                        (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))),
                "1:Selenium, Custom, " +
                        "2:TestNG, JUnit, Custom, " +
                        "3:TestNG, JUnit, Custom, " +
                        "4:Log4J, TestNG log, Custom, " +
                        "5:Jenkins, Allure, Custom, " +
                        "6:Custom");
    }

    @Test
    public void rowByCriteriaIntTest() {
        MapArray<String, ICell> row = table().row("MSTest, NUnit, Epam", column(3));
        areEquals(print(row.select(
                        (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))),
                "Type:Test Runner, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam");
    }

    @Test
    public void rowByCriteriaStringTest() {
        MapArray<String, ICell> row = table().row("MSTest, NUnit, Epam", column("Plans"));
        areEquals(print(row.select(
                        (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))),
                "Type:Test Runner, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam");
    }

    @Test
    public void rowsByCriteriaTest() {
        MapArray<String, MapArray<String, ICell>> rows = table().rows("Plans=MSTest, NUnit, Epam");
        new Check("Rows count").areEquals(rows.size(), 2);
        new Check("Rows content").areEquals(print(rows.select(
                        (k, v) -> format("%s:%s", k, v.select(
                                (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))))),
                "2:[Type:Test Runner, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam], " +
                        "3:[Type:Asserter, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam]");
    }

    @Test
    public void rowsByTwoCriteriasTest() {
        MapArray<String, MapArray<String, ICell>> rows = table().rows("Plans=MSTest, NUnit, Epam", "Type=Asserter");
        new Check("Rows count").areEquals(rows.size(), 1);
        new Check("Rows content").areEquals(print(rows.select(
                        (k, v) -> format("%s:%s", k, v.select(
                                (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))))),
                "3:[Type:Asserter, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam]");
    }

    @Test
    public void columnsByCriteriaTest() {
        MapArray<String, MapArray<String, ICell>> columns = table().columns("1=Selenium, Custom");
        new Check("Columns count").areEquals(columns.size(), 1);
        new Check("Columns content").areEquals(print(columns.select(
                        (k, v) -> format("%s:%s", k, v.select(
                                (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))))),
                "Now:[1:Selenium, Custom, " +
                        "2:TestNG, JUnit, Custom, " +
                        "3:TestNG, JUnit, Custom, " +
                        "4:Log4J, TestNG log, Custom, " +
                        "5:Jenkins, Allure, Custom, " +
                        "6:Custom]");
    }

    @Test
    public void columnsByTwoCriteriasTest() {
        MapArray<String, MapArray<String, ICell>> columns = table().columns("2=Test Runner", "4=Logger");
        new Check("Columns count").areEquals(columns.size(), 1);
        new Check("Columns content").areEquals(print(columns.select(
                        (k, v) -> format("%s:%s", k, v.select(
                                (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))))),
                "Type:[1:Drivers, " +
                        "2:Test Runner, " +
                        "3:Asserter, " +
                        "4:Logger, " +
                        "5:Reporter, " +
                        "6:BDD/DSL]");
    }

    @Test
    public void columnsGetTest() {
        MapArray<String, MapArray<String, ICell>> columns = table().columns().get();
        new Check("Columns count").areEquals(columns.size(), 3);

        new Check("Columns content").areEquals(print(columns.select(
                        (k, v) -> format("%s:%s", k, v.select(
                                (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))))),
                "Type:[1:Drivers, " +
                        "2:Test Runner, " +
                        "3:Asserter, " +
                        "4:Logger, " +
                        "5:Reporter, " +
                        "6:BDD/DSL], " +
                        "Now:[1:Selenium, Custom, " +
                        "2:TestNG, JUnit, Custom, " +
                        "3:TestNG, JUnit, Custom, " +
                        "4:Log4J, TestNG log, Custom, " +
                        "5:Jenkins, Allure, Custom, " +
                        "6:Custom], " +
                        "Plans:[1:JavaScript, Appium, WinAPI, Sikuli, " +
                        "2:MSTest, NUnit, Epam, " +
                        "3:MSTest, NUnit, Epam, " +
                        "4:Epam, XML/Json logging, Hyper logging, " +
                        "5:EPAM Report portal, Serenity, TimCity, Hudson, " +
                        "6:Cucumber, Jbehave, Thucydides, SpecFlow]");
    }

    @Test
    public void columnsGetAsTextTest() {
        MapArray<String, MapArray<String, String>> columns = table().columns().getAsText();
        new Check("Columns count").areEquals(columns.size(), 3);
        new Check("Columns content").areEquals(print(columns.select(
                        (k, v) -> format("%s:%s", k, v.select(
                                (rowK, rowV) -> format("%s:%s", rowK, rowV))))),
                "Type:[1:Drivers, " +
                        "2:Test Runner, " +
                        "3:Asserter, " +
                        "4:Logger, " +
                        "5:Reporter, " +
                        "6:BDD/DSL], " +
                        "Now:[1:Selenium, Custom, " +
                        "2:TestNG, JUnit, Custom, " +
                        "3:TestNG, JUnit, Custom, " +
                        "4:Log4J, TestNG log, Custom, " +
                        "5:Jenkins, Allure, Custom, " +
                        "6:Custom], " +
                        "Plans:[1:JavaScript, Appium, WinAPI, Sikuli, " +
                        "2:MSTest, NUnit, Epam, " +
                        "3:MSTest, NUnit, Epam, " +
                        "4:Epam, XML/Json logging, Hyper logging, " +
                        "5:EPAM Report portal, Serenity, TimCity, Hudson, " +
                        "6:Cucumber, Jbehave, Thucydides, SpecFlow]");
    }

    @Test
    public void columnsGetByNumTest() {
        MapArray<String, ICell> column = table().columns().getColumn(2);
        new Check("Column content").areEquals(print(column.select(
                (rowK, rowV) -> format("%s:%s", rowK, rowV.getValue()))), expectedColumn);
    }

    @Test
    public void columnsGetByNameTest() {
        MapArray<String, ICell> column = table().columns().getColumn("Now");
        new Check("Column content").areEquals(print(column.select(
                (rowK, rowV) -> format("%s:%s", rowK, rowV.getValue()))), expectedColumn);
    }

    @Test
    public void columnsGetByNumAsTextTest() {
        MapArray<String, String> column = table().columns().getColumnAsText(2);
        new Check("Column content").areEquals(print(column.select(
                (rowK, rowV) -> format("%s:%s", rowK, rowV))), expectedColumn);
    }

    @Test
    public void columnsGetByNameAsTextTest() {
        MapArray<String, String> column = table().columns().getColumnAsText("Now");
        new Check("Column content").areEquals(print(column.select(
                (rowK, rowV) -> format("%s:%s", rowK, rowV))), expectedColumn);
    }

    @Test
    public void rowsGetTest() {
        MapArray<String, MapArray<String, ICell>> rows = table().rows().get();
        new Check("Rows count").areEquals(rows.size(), 6);
        new Check("Rows content").areEquals(print(rows.select(
                        (k, v) -> format("%s:%s", k, v.select(
                                (rowK, rowV) -> format("%s:%s", rowK, rowV.getText()))))),
                "1:[Type:Drivers, " +
                        "Now:Selenium, Custom, " +
                        "Plans:JavaScript, Appium, WinAPI, Sikuli], " +
                        "2:[Type:Test Runner, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam], " +
                        "3:[Type:Asserter, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam], " +
                        "4:[Type:Logger, " +
                        "Now:Log4J, TestNG log, Custom, " +
                        "Plans:Epam, XML/Json logging, Hyper logging], " +
                        "5:[Type:Reporter, " +
                        "Now:Jenkins, Allure, Custom, " +
                        "Plans:EPAM Report portal, Serenity, TimCity, Hudson], " +
                        "6:[Type:BDD/DSL, " +
                        "Now:Custom, " +
                        "Plans:Cucumber, Jbehave, Thucydides, SpecFlow]");
    }

    @Test
    public void rowsGetAsTextTest() {
        MapArray<String, MapArray<String, String>> rows = table().rows().getAsText();
        new Check("Rows count").areEquals(rows.size(), 6);
        new Check("Rows content").areEquals(print(rows.select(
                        (k, v) -> format("%s:%s", k, v.select(
                                (rowK, rowV) -> format("%s:%s", rowK, rowV))))),
                "1:[Type:Drivers, " +
                        "Now:Selenium, Custom, " +
                        "Plans:JavaScript, Appium, WinAPI, Sikuli], " +
                        "2:[Type:Test Runner, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam], " +
                        "3:[Type:Asserter, " +
                        "Now:TestNG, JUnit, Custom, " +
                        "Plans:MSTest, NUnit, Epam], " +
                        "4:[Type:Logger, " +
                        "Now:Log4J, TestNG log, Custom, " +
                        "Plans:Epam, XML/Json logging, Hyper logging], " +
                        "5:[Type:Reporter, " +
                        "Now:Jenkins, Allure, Custom, " +
                        "Plans:EPAM Report portal, Serenity, TimCity, Hudson], " +
                        "6:[Type:BDD/DSL, " +
                        "Now:Custom, " +
                        "Plans:Cucumber, Jbehave, Thucydides, SpecFlow]");
    }

    @Test
    public void rowsGetByNumTest() {
        MapArray<String, ICell> row = table().rows().getRow(3);
        new Check("Row content").areEquals(print(row.select(
                (rowK, rowV) -> format("%s:%s", rowK, rowV.getValue()))), expectedRow);
    }

    @Test
    public void rowsGetByNameTest() {
        MapArray<String, ICell> row = table().rows().getRow("3");
        new Check("Row content").areEquals(print(row.select(
                (rowK, rowV) -> format("%s:%s", rowK, rowV.getValue()))), expectedRow);
    }

    @Test
    public void rowsGetByNumAsTextTest() {
        MapArray<String, String> row = table().rows().getRowAsText(3);
        new Check("Row content").areEquals(print(row.select(
                (rowK, rowV) -> format("%s:%s", rowK, rowV))), expectedRow);
    }

    @Test
    public void rowsGetByNameAsTextTest() {
        MapArray<String, String> row = table().rows().getRowAsText("3");
        new Check("Row content").areEquals(print(row.select(
                (rowK, rowV) -> format("%s:%s", rowK, rowV))), expectedRow);
    }

    @Test
    public void cellsToColumnTest(){
        MapArray<String, ICell> cellsToColumn = table().columns().cellsToColumn(Arrays.asList(table().cell(1, 1), table().cell(2, 2)));

        new Check("Columns Headers").areEquals("Type, Now", cellsToColumn.key(0)+", "+cellsToColumn.key(1));
        new Check("Cells Values").areEquals("Drivers, TestNG, JUnit, Custom", print(new String[]{cellsToColumn.value(0).getValue(), cellsToColumn.value(1).getValue()}));
    }

    @Test
    public void cellsToRowTest(){
        MapArray<String, ICell> cellsToRow = table().rows().cellsToRow(Arrays.asList(table().cell(1, 1), table().cell(2, 2)));

        new Check("Rows Indexes").areEquals("1, 2", cellsToRow.key(0)+", "+cellsToRow.key(1));
        new Check("Cells Values").areEquals("Drivers, TestNG, JUnit, Custom", print(new String[]{cellsToRow.value(0).getValue(), cellsToRow.value(1).getValue()}));
    }

    @Test
    public void rowValueByNameTests() {
        new Check("Row Value").areEquals(table().rowValue("2"), Arrays.asList("Test Runner", "TestNG, JUnit, Custom", "MSTest, NUnit, Epam"));
    }

    @Test
    public void rowValueByNumberTests() {
        new Check("Row Value").areEquals(table().rowValue(2), Arrays.asList("Test Runner", "TestNG, JUnit, Custom", "MSTest, NUnit, Epam"));
    }

    @Test
    public void columnValueByNameTests() {
        new Check("Column Value").areEquals(table().columnValue("Now"), Arrays.asList("Selenium, Custom", "TestNG, JUnit, Custom", "TestNG, JUnit, Custom", "Log4J, TestNG log, Custom", "Jenkins, Allure, Custom", "Custom"));
    }

    @Test
    public void columnValueByNumberTests() {
        new Check("Column Value").areEquals(table().columnValue(2), Arrays.asList("Selenium, Custom", "TestNG, JUnit, Custom", "TestNG, JUnit, Custom", "Log4J, TestNG log, Custom", "Jenkins, Allure, Custom", "Custom"));
    }

}

File: C#.Net/Tests/JDI UIWebTests/JDI UIWebTests/Tests/Complex/Table/SearchRowsColumnsTests.cs
using System.Linq;
using JDI_Commons;
using JDI_Matchers.NUnit;
using JDI_UIWebTests.Tests.Complex.Table.Base;
using JDI_Web.Selenium.Elements.Complex.Table;
using NUnit.Framework;

namespace JDI_UIWebTests.Tests.Complex.Table
{
    [TestFixture]
    public class SearchRowsColumnsTests : SupportTableTestBase
    {
        private static readonly string _expectedColumn2 =
            "1:Selenium, Custom, " +
            "2:TestNG, JUnit, Custom, " +
            "3:TestNG, JUnit, Custom, " +
            "4:Log4J, TestNG log, Custom, " +
            "5:Jenkins, Allure, Custom, " +
            "6:Custom";

        private static readonly string _expectedRow2 =
            "Type:Test Runner, " +
            "Now:TestNG, JUnit, Custom, " +
            "Plans:MSTest, NUnit, Epam";

        [Test]
        public void ColumnByNumTest()
        {
            var column = Table.Column(2).Select(pair => $"{pair.Key}:{pair.Value.GetText}");
            new Check().AreEquals(column.Print(), _expectedColumn2);
        }

        [Test]
        public void ColumnByNameTest()
        {
            var column = Table.Column("Now").Select(pair => $"{pair.Key}:{pair.Value.GetText}");
            new Check().AreEquals(column.Print(), _expectedColumn2);
        }

        [Test]
        public void RowByNumTest()
        {
            var row = Table.Row(2).Select(pair => $"{pair.Key}:{pair.Value.GetText}");
            new Check().AreEquals(row.Print(), _expectedRow2);
        }

        [Test]
        public void RowByNameTest()
        {
            var row = Table.Row("2").Select(pair => $"{pair.Key}:{pair.Value.GetText}");
            new Check().AreEquals(row.Print(), _expectedRow2);
        }

        [Test]
        public void ColumnByCriteriaIntTest()
        {
            var column = Table.Column("TestNG, JUnit, Custom", Row.row(3)).Select(pair => $"{pair.Key}:{pair.Value.GetText}");
            new Check().AreEquals(column.Print(), _expectedColumn2);
        }

        [Test]
        public void ColumnByCriteriaStringTest()
        {
            var column = Table.Column("TestNG, JUnit, Custom", Row.row("3")).Select(pair => $"{pair.Key}:{pair.Value.GetText}");
            new Check().AreEquals(column.Print(), _expectedColumn2);
        }

        [Test]
        public void RowByCriteriaIntTest()
        {
            var row = Table.Row("MSTest, NUnit, Epam", Column.column(3)).Select(pair => $"{pair.Key}:{pair.Value.GetText}");
            new Check().AreEquals(row.Print(), _expectedRow2);
        }

        [Test]
        public void RowByCriteriaStringTest()
        {
            var row = Table.Row("MSTest, NUnit, Epam", Column.column("Plans")).Select(pair => $"{pair.Key}:{pair.Value.GetText}");
            new Check().AreEquals(row.Print(), _expectedRow2);
        }

        [Test]
        public void RowsByCriteriaTest()
        {
            var rows = Table.GetRows("Plans=MSTest, NUnit, Epam");
            new Check("Rows count").AreEquals(rows.Count, 2);
            string stringRow =
                rows.Select(
                    pair => $"{pair.Key}:[{pair.Value.Select(pair2 => $"{pair2.Key}:{pair2.Value.GetText}").Print()}]")
                    .Print();
            new Check("Rows content").AreEquals(
                stringRow,
                "2:[Type:Test Runner, Now:TestNG, JUnit, Custom, Plans:MSTest, NUnit, Epam], " +
                "3:[Type:Asserter, Now:TestNG, JUnit, Custom, Plans:MSTest, NUnit, Epam]");
        }

        [Test]
        public void RowsByTwoCriteriasTest()
        {
            var rows = Table.GetRows("Plans=MSTest, NUnit, Epam", "Type=Asserter");
            new Check("Rows count").AreEquals(rows.Count, 1);
            string stringRow =
                rows.Select(
                    pair => $"{pair.Key}:[{pair.Value.Select(pair2 => $"{pair2.Key}:{pair2.Value.GetText}").Print()}]")
                    .Print();
            new Check("Rows content").AreEquals(
                stringRow,
                "3:[Type:Asserter, Now:TestNG, JUnit, Custom, Plans:MSTest, NUnit, Epam]");
        }

        [Test]
        public void ColumnsByCriteriaTest()
        {
            var columns = Table.GetColumns("1=Selenium, Custom");
            new Check().AreEquals(columns.Count, 1);
            string stringColumns =
                columns.Select(
                    pair => $"{pair.Key}:[{pair.Value.Select(pair2 => $"{pair2.Key}:{pair2.Value.GetText}").Print()}]")
                    .Print();
            new Check("Columns content").AreEquals(
                stringColumns,
                "Now:[1:Selenium, Custom, " +
                "2:TestNG, JUnit, Custom, " +
                "3:TestNG, JUnit, Custom, " +
                "4:Log4J, TestNG log, Custom, " +
                "5:Jenkins, Allure, Custom, " +
                "6:Custom]");
        }

        [Test]
        public void ColumnsByTwoCriteriasTest()
        {
            var columns = Table.GetColumns("2=Test Runner", "4=Logger");
            new Check().AreEquals(columns.Count, 1);
            string stringColumns =
                columns.Select(
                    pair => $"{pair.Key}:[{pair.Value.Select(pair2 => $"{pair2.Key}:{pair2.Value.GetText}").Print()}]")
                    .Print();
            new Check("Columns content").AreEquals(
                stringColumns,
                "Type:[1:Drivers, " +
                "2:Test Runner, " +
                "3:Asserter, " +
                "4:Logger, " +
                "5:Reporter, " +
                "6:BDD/DSL]");
        }

        [Test]
        public void ColumnsGetTest()
        {
            var columns = Table.Columns.Get();
            new Check("Columns count").AreEquals(columns.Count, 3);
            string stringColumns = columns.Select(
                pair => $"{pair.Key}:[{pair.Value.Select(pair2 => $"{pair2.Key}:{pair2.Value.GetText}").Print()}]")
                .Print();
            new Check("Columns content").AreEquals(
                stringColumns,
                "Type:[1:Drivers, " +
                "2:Test Runner, " +
                "3:Asserter, " +
                "4:Logger, " +
                "5:Reporter, " +
                "6:BDD/DSL], " +
                "Now:[1:Selenium, Custom, " +
                "2:TestNG, JUnit, Custom, " +
                "3:TestNG, JUnit, Custom, " +
                "4:Log4J, TestNG log, Custom, " +
                "5:Jenkins, Allure, Custom, " +
                "6:Custom], " +
                "Plans:[1:JavaScript, Appium, WinAPI, Sikuli, " +
                "2:MSTest, NUnit, Epam, " +
                "3:MSTest, NUnit, Epam, " +
                "4:Epam, XML/Json logging, Hyper logging, " +
                "5:EPAM Report portal, Serenity, TimCity, Hudson, " +
                "6:Cucumber, Jbehave, Thucydides, SpecFlow]");
        }

        [Test]
        public void ColumnsAsTextTest()
        {
            var columns = Table.Columns.AsText;
            new Check("Columns count").AreEquals(columns.Count, 3);
            string stringColumns = columns.Select(
                pair => $"{pair.Key}:[{pair.Value.Select(pair2 => $"{pair2.Key}:{pair2.Value}").Print()}]")
                .Print();
            new Check("Columns content").AreEquals(
                stringColumns,
                "Type:[1:Drivers, " +
                "2:Test Runner, " +
                "3:Asserter, " +
                "4:Logger, " +
                "5:Reporter, " +
                "6:BDD/DSL], " +
                "Now:[1:Selenium, Custom, " +
                "2:TestNG, JUnit, Custom, " +
                "3:TestNG, JUnit, Custom, " +
                "4:Log4J, TestNG log, Custom, " +
                "5:Jenkins, Allure, Custom, " +
                "6:Custom], " +
                "Plans:[1:JavaScript, Appium, WinAPI, Sikuli, " +
                "2:MSTest, NUnit, Epam, " +
                "3:MSTest, NUnit, Epam, " +
                "4:Epam, XML/Json logging, Hyper logging, " +
                "5:EPAM Report portal, Serenity, TimCity, Hudson, " +
                "6:Cucumber, Jbehave, Thucydides, SpecFlow]");
        }

        [Test]
        public void ColumnsGetByNumTest()
        {
            var column = Table.Columns.GetColumn(2);
            new Check("Column content").AreEquals(
                column.Select(pair => $"{pair.Key}:{pair.Value.Value}").Print(),
                _expectedColumn2);
        }

        [Test]
        public void ColumnsGetByNameTest()
        {
            var column = Table.Columns.GetColumn("Now");
            new Check("Column content").AreEquals(
                column.Select(pair => $"{pair.Key}:{pair.Value.Value}").Print(),
                _expectedColumn2);
        }

        [Test]
        public void ColumnsGetByNumAsTextTest()
        {
            var column = Table.Columns.GetColumnAsText(2);
            new Check("Column content").AreEquals(
                column.Select(pair => $"{pair.Key}:{pair.Value}").Print(),
                _expectedColumn2);
        }

        [Test]
        public void ColumnsGetByNameAsTextTest()
        {
            var column = Table.Columns.GetColumnAsText("Now");
            new Check("Column content").AreEquals(
                column.Select(pair => $"{pair.Key}:{pair.Value}").Print(),
                _expectedColumn2);
        }

        [Test]
        public void RowsGetTest()
        {
            var rows = Table.Rows.Get();
            new Check("Rows count").AreEquals(rows.Count, 6);
            new Check("Rows content").AreEquals(
                rows.Select(
                    pair => $"{pair.Key}:[{pair.Value.Select(pair2 => $"{pair2.Key}:{pair2.Value.GetText}").Print()}]")
                    .Print(),
                "1:[Type:Drivers, " +
                "Now:Selenium, Custom, " +
                "Plans:JavaScript, Appium, WinAPI, Sikuli], " +
                "2:[Type:Test Runner, " +
                "Now:TestNG, JUnit, Custom, " +
                "Plans:MSTest, NUnit, Epam], " +
                "3:[Type:Asserter, " +
                "Now:TestNG, JUnit, Custom, " +
                "Plans:MSTest, NUnit, Epam], " +
                "4:[Type:Logger, " +
                "Now:Log4J, TestNG log, Custom, " +
                "Plans:Epam, XML/Json logging, Hyper logging], " +
                "5:[Type:Reporter, " +
                "Now:Jenkins, Allure, Custom, " +
                "Plans:EPAM Report portal, Serenity, TimCity, Hudson], " +
                "6:[Type:BDD/DSL, " +
                "Now:Custom, " +
                "Plans:Cucumber, Jbehave, Thucydides, SpecFlow]");
        }

        [Test]
        public void RowsGetAsTextTest()
        {
            var rows = Table.Rows.AsText;
            new Check("Rows count").AreEquals(rows.Count, 6);
            new Check("Rows content").AreEquals(
                rows.Select(
                    pair => $"{pair.Key}:[{pair.Value.Select(pair2 => $"{pair2.Key}:{pair2.Value}").Print()}]")
                    .Print(),
                "1:[Type:Drivers, " +
                "Now:Selenium, Custom, " +
                "Plans:JavaScript, Appium, WinAPI, Sikuli], " +
                "2:[Type:Test Runner, " +
                "Now:TestNG, JUnit, Custom, " +
                "Plans:MSTest, NUnit, Epam], " +
                "3:[Type:Asserter, " +
                "Now:TestNG, JUnit, Custom, " +
                "Plans:MSTest, NUnit, Epam], " +
                "4:[Type:Logger, " +
                "Now:Log4J, TestNG log, Custom, " +
                "Plans:Epam, XML/Json logging, Hyper logging], " +
                "5:[Type:Reporter, " +
                "Now:Jenkins, Allure, Custom, " +
                "Plans:EPAM Report portal, Serenity, TimCity, Hudson], " +
                "6:[Type:BDD/DSL, " +
                "Now:Custom, " +
                "Plans:Cucumber, Jbehave, Thucydides, SpecFlow]");
        }

        [Test]
        public void RowsGetByNumTest()
        {
            var row = Table.Rows.GetRow(3);
            new Check("Row content").AreEquals(
                row.Select(pair => $"{pair.Key}:{pair.Value.Value}").Print(),
                "Type:Asserter, Now:TestNG, JUnit, Custom, Plans:MSTest, NUnit, Epam");
        }

        [Test]
        public void RowsGetByNameTest()
        {
            var row = Table.Rows.GetRow("3");
            new Check("Row content").AreEquals(
                row.Select(pair => $"{pair.Key}:{pair.Value.Value}").Print(),
                "Type:Asserter, Now:TestNG, JUnit, Custom, Plans:MSTest, NUnit, Epam");
        }

        [Test]
        public void RowsGetByNumAsTextTest()
        {
            var row = Table.Rows.GetRowAsText(3);
            new Check("Row content").AreEquals(
                row.Select(pair => $"{pair.Key}:{pair.Value}").Print(),
                "Type:Asserter, Now:TestNG, JUnit, Custom, Plans:MSTest, NUnit, Epam");
        }

        [Test]
        public void RowsGetByNameAsTextTest()
        {
            var row = Table.Rows.GetRowAsText("3");
            new Check("Row content").AreEquals(
                row.Select(pair => $"{pair.Key}:{pair.Value}").Print(),
                "Type:Asserter, Now:TestNG, JUnit, Custom, Plans:MSTest, NUnit, Epam");
        }

        [Test]
        public void CellsToColumnTest()
        {
            var cells = Table.Columns.CellsToColumn(new[] { Table.Cell(1, 1), Table.Cell(2, 2) });
            new Check("Columns Header").CollectionEquals(new[] { "Type", "Now" }, cells.Select(pair => pair.Key));
            new Check("Cells values").CollectionEquals(new[] { "Drivers", "TestNG, JUnit, Custom" },
                cells.Select(pair => pair.Value.Value));
        }

        [Test]
        public void CellsToRowTest()
        {
            var rows = Table.Rows.CellsToRow(new[] { Table.Cell(1, 1), Table.Cell(2, 2) });
            new Check("Rows Indexes").CollectionEquals(new[] { "1", "2" }, rows.Select(pair => pair.Key));
            new Check("Cells values").CollectionEquals(new[] { "Drivers", "TestNG, JUnit, Custom" },
                rows.Select(pair => pair.Value.Value));
        }

        [Test]
        public void RowValueByNumTests()
        {
            new Check("Row value").CollectionEquals(Table.RowValue(2),
                new[] { "Test Runner", "TestNG, JUnit, Custom", "MSTest, NUnit, Epam" });
        }

        [Test]
        public void RowValueByNameTests()
        {
            new Check("Row value").CollectionEquals(Table.RowValue("2"),
                new[] { "Test Runner", "TestNG, JUnit, Custom", "MSTest, NUnit, Epam" });
        }

        [Test]
        public void ColumnValueByNumTests()
        {
            new Check("Column value").CollectionEquals(Table.ColumnValue(2), new[] { "Selenium, Custom", "TestNG, JUnit, Custom", "TestNG, JUnit, Custom", "Log4J, TestNG log, Custom", "Jenkins, Allure, Custom", "Custom" });
        }

        [Test]
        public void ColumnValueByNameTests()
        {
            new Check("Column value").CollectionEquals(Table.ColumnValue("Now"), new[] { "Selenium, Custom", "TestNG, JUnit, Custom", "TestNG, JUnit, Custom", "Log4J, TestNG log, Custom", "Jenkins, Allure, Custom", "Custom" });
        }
    }
}
